src/main/java/cloud/prefab/client/config/ConfigResolver.java

```diff
@@ -1,221 +1,371 @@
+
+          return new EvaluatedCriterion(criterion, prop.get(), matched);
+        } else {
+          return new EvaluatedCriterion(criterion, false);
+        }
+      case PROP_DOES_NOT_END_WITH_ONE_OF:
+        if (prop.isPresent() && prop.get().hasString()) {
+          final boolean matched = criterion
+            .getValueToMatch()
+            .getStringList()
+            .getValuesList()
+            .stream()
+            .anyMatch(value -> prop.get().getString().endsWith(value));
+
+          return new EvaluatedCriterion(criterion, prop.get(), !matched);
+        } else {
+          return new EvaluatedCriterion(criterion, true);
+        }
+    }
+    // Unknown Operator
+    return new EvaluatedCriterion(criterion, false);
   }
 
-  NamespaceMatch evaluateMatch(String namespace, String baseNamespace) {
-    final String[] nsSplit = namespace.split(NAMESPACE_DELIMITER);
-    final String[] baseSplit = baseNamespace.split(NAMESPACE_DELIMITER);
+  /**
+   * a.b.c match a.b -> true
+   * a.b match a.b.c -> false
+   *
+   * @param valueToMatch
+   * @param propertyString
+   * @return
+   */
+  boolean hierarchicalMatch(String propertyString, String valueToMatch) {
+    return propertyString.startsWith(valueToMatch);
+  }
 
-    List<Boolean> matches = new ArrayList<>();
-    for (int i = 0; i < nsSplit.length; i++) {
-      if (baseSplit.length <= i) {
-        matches.add(false);
-        continue;
-      }
-      matches.add(nsSplit[i].equals(baseSplit[i]));
+  public boolean setProjectEnvId(Prefab.Configs configs) {
+    if (configs.hasConfigServicePointer()) {
+      this.projectEnvId = configs.getConfigServicePointer().getProjectEnvId();
+      return true;
     }
-    return new NamespaceMatch(
-      matches.stream().allMatch(b -> b),
-      matches.stream().filter(b -> b).count()
-    );
+    return false;
   }
 
   public Collection<String> getKeys() {
-    return localMap.get().keySet();
+    return configStore.getKeys();
   }
 
   public String contentsString() {
     StringBuilder sb = new StringBuilder("\n");
-    List<String> sortedKeys = new ArrayList(localMap.get().keySet());
+    List<String> sortedKeys = new ArrayList(getKeys());
     Collections.sort(sortedKeys);
     for (String key : sortedKeys) {
-      ResolverElement resolverElement = localMap.get().get(key);
-      sb.append(padded(key, 30));
-      sb.append(padded(toS(resolverElement.getConfigValue()), 40));
-      sb.append(padded(resolverElement.provenance(), 90));
+      ConfigElement configElement = configStore.getElement(key);
+      final Optional<Match> match = findMatch(configElement, new HashMap<>());
+
+      if (match.isPresent()) {
+        sb.append(padded(key, 30));
+        sb.append(padded(toS(match.get().getConfigValue()), 40));
+        sb.append(padded(configElement.getProvenance().toString(), 40));
+        sb.append(padded(match.get().getReason(), 40));
+      }
       sb.append("\n");
     }
     System.out.println(sb.toString());
```

Comment:
is this println supposed to be here?

Reply:
yeah, let's put this on the list to improve. Should be a log message from "cloud.prefab.config.initialization" or something.